test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java

```diff
@@ -0,0 +1,401 @@
+                        builder.requireClusterFeature(parser.text());
+                    }
+                }
+            }
+        }
+        parser.nextToken();
+    }
+
+    public static final PrerequisiteSection EMPTY = new PrerequisiteSection();
+
+    private final List<Predicate<ClientYamlTestExecutionContext>> skipCriteriaList;
+    private final List<Predicate<ClientYamlTestExecutionContext>> requireCriteriaList;
+    private final List<String> yamlRunnerFeatures;
+    final String skipReason;
+    final String requireReason;
+
+    private PrerequisiteSection() {
+        this.skipCriteriaList = new ArrayList<>();
+        this.requireCriteriaList = new ArrayList<>();
+        this.yamlRunnerFeatures = new ArrayList<>();
+        this.skipReason = null;
+        this.requireReason = null;
+    }
+
+    PrerequisiteSection(
+        List<Predicate<ClientYamlTestExecutionContext>> skipCriteriaList,
+        String skipReason,
+        List<Predicate<ClientYamlTestExecutionContext>> requireCriteriaList,
+        String requireReason,
+        List<String> yamlRunnerFeatures
+    ) {
+        this.skipCriteriaList = skipCriteriaList;
+        this.requireCriteriaList = requireCriteriaList;
+        this.yamlRunnerFeatures = yamlRunnerFeatures;
+        this.skipReason = skipReason;
+        this.requireReason = requireReason;
+    }
+
+    public boolean yamlRunnerHasFeature(String feature) {
+        return yamlRunnerFeatures.contains(feature);
+    }
+
+    boolean skipCriteriaMet(ClientYamlTestExecutionContext context) {
+        return skipCriteriaList.stream().anyMatch(c -> c.test(context));
+    }
+
+    boolean requiresCriteriaMet(ClientYamlTestExecutionContext context) {
+        return requireCriteriaList.stream().allMatch(c -> c.test(context));
+    }
+
+    public void evaluate(ClientYamlTestExecutionContext context, String testCandidateDescription) {
+        if (isEmpty()) {
+            return;
+        }
+
+        if (requiresCriteriaMet(context) == false) {
+            throw new AssumptionViolatedException(buildMessage(testCandidateDescription, false));
+        }
+
+        if (skipCriteriaMet(context)) {
+            throw new AssumptionViolatedException(buildMessage(testCandidateDescription, true));
+        }
+    }
+
+    boolean isEmpty() {
+        return skipCriteriaList.isEmpty() && requireCriteriaList.isEmpty() && yamlRunnerFeatures.isEmpty();
+    }
+
+    String buildMessage(String description, boolean isSkip) {
+        StringBuilder messageBuilder = new StringBuilder();
+        messageBuilder.append("[").append(description).append("] skipped,");
+        var reason = isSkip ? skipReason : requireReason;
+        if (Strings.isNullOrEmpty(reason) == false) {
+            messageBuilder.append(" reason: [").append(reason).append("]");
+        }
+        if (yamlRunnerFeatures.isEmpty() == false) {
+            messageBuilder.append(" unsupported features ").append(yamlRunnerFeatures);
```

Comment:
Just a thought, how about appending that to the `requiresReason` in the builder when required runner features are missing?

Reply:
It's a good idea, but I prefer to leave it as is for now, so the output message does not change too much.